Runtime error when calling DeleteAllById(Iterable<T> list) if the repository extends DynamicProxyable

In the case of an **application repository extending ReactiveCouchbaseRepository and DynamicProxyable** interfaces, calling **deleteAllById(Iterable<? extends ID> ids)** method on the application repository will result in the following **runtime exception**:

`
java.lang.reflect.UndeclaredThrowableException
	at jdk.proxy2/jdk.proxy2.$Proxy126.deleteAllById(Unknown Source)
	at org.springframework.data.couchbase.repository.query.ReactiveCouchbaseRepositoryQueryCollectionIntegrationTests.myTest1(ReactiveCouchbaseRepositoryQueryCollectionIntegrationTests.java:128)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: java.lang.NoSuchMethodException: No such method: deleteAllById
	at org.springframework.data.couchbase.repository.support.FindMethod.internalFind(FindMethod.java:84)
	at org.springframework.data.couchbase.repository.support.FindMethod.findMethod(FindMethod.java:33)
	at org.springframework.data.couchbase.repository.support.DynamicInvocationHandler.invoke(DynamicInvocationHandler.java:118)
	... 5 more
`

Test class: **ReactiveCouchbaseRepositoryQueryCollectionIntegrationTests**
Test method to reproduce:

```
@Test
	public void deleteAllByIdDynamicProxyableTest() {

		ReactiveAirportRepository ar = reactiveAirportRepository.withScope(scopeName).withCollection(collectionName);
		Airport vie = new Airport("airports::vie", "vie", "loww");
		try {
			Airport saved = ar.save(vie).block();
			ar.deleteAllById(List.of(saved.getId())).block();
		} catch (Exception e) {
			e.printStackTrace();
			throw e;
		} finally {
			ar.delete(vie).block();
		}
	}
```
The problem does not exhibit if the application repository extends only ReactiveCouchbaseRepository.

Test class: **ReactiveCouchbaseRepositoryKeyValueIntegrationTests**

Test method that passes:
```
@Test
    @IgnoreWhen(clusterTypes = ClusterType.MOCKED)
    void deleteAllByIdNoDynamicProxyableTest() {
        // the User class has a version.
        User user = new User(UUID.randomUUID().toString(), "f", "l");
        // save the document - we don't care how on this call
        userRepository.save(user).block();
        // Now set the version to 0, it should attempt an insert and fail.
        long saveVersion = user.getVersion();
        user.setVersion(0);
        assertThrows(DuplicateKeyException.class, () -> userRepository.save(user).block());
        user.setVersion(saveVersion + 1);
        assertThrows(OptimisticLockingFailureException.class, () -> userRepository.save(user).block());
        userRepository.delete(user);

        // Airline does not have a version
        Airline airline = new Airline(UUID.randomUUID().toString(), "MyAirline", null);
        // save the document - we don't care how on this call
        reactiveAirlineRepository.save(airline).block();
        reactiveAirlineRepository.save(airline).block(); // If it was an insert it would fail. Can't tell if an upsert or
        // cleanup.
        reactiveAirlineRepository.deleteAllById(List.of(airline.getId())).block();
    }
```

Tested with **spring data couchbase 5.2.3, couchbase client 2.4.11**

--- a/src/main/java/org/springframework/data/couchbase/repository/support/DynamicInvocationHandler.java
+++ b/src/main/java/org/springframework/data/couchbase/repository/support/DynamicInvocationHandler.java
@@ -110,7 +110,8 @@ public class DynamicInvocationHandler<T> implements InvocationHandler {
 					.map(o -> o == null ? null : (o.getClass() == entityInformation.getJavaType() ? Object.class : o.getClass()))
 					.toArray(Class<?>[]::new);
 			// the CouchbaseRepository methods - findById(id) etc - will have a parameter type of Object instead of ID
-			if (method.getName().endsWith("ById") && args.length == 1) {
+			// but deleteByAllId first parameter will be an iterable.
+			if (method.getName().endsWith("ById") && args.length == 1 && ! Iterable.class.isAssignableFrom(paramTypes[0]) ) {
 				paramTypes[0] = Object.class;
 			}
 		}

--- a/src/main/java/org/springframework/data/couchbase/repository/support/SimpleCouchbaseRepository.java
+++ b/src/main/java/org/springframework/data/couchbase/repository/support/SimpleCouchbaseRepository.java
@@ -73,7 +73,7 @@ public class SimpleCouchbaseRepository<T, ID> extends CouchbaseRepositoryBase<T,
 		String scopeName = getScope();
 		String collectionName = getCollection();
 		// clear out the PseudoArgs here as whatever is called by operations.save() could be in a different thread.
-		// not that this will also clear out Options, but that's ok as any options would not work
+		// note that this will also clear out Options, but that's ok as any options would not work
 		// with all of insert/upsert/replace. If Options are needed, use template.insertById/upsertById/replaceById
 		getReactiveTemplate().setPseudoArgs(null);
 		return operations.save(entity, scopeName, collectionName);
@@ -82,7 +82,13 @@ public class SimpleCouchbaseRepository<T, ID> extends CouchbaseRepositoryBase<T,
 	@Override
 	public <S extends T> Iterable<S> saveAll(Iterable<S> entities) {
 		Assert.notNull(entities, "The given Iterable of entities must not be null!");
-		return Streamable.of(entities).stream().map((e) -> save(e)).collect(StreamUtils.toUnmodifiableList());
+		String scopeName = getScope();
+		String collectionName = getCollection();
+		// clear out the PseudoArgs here as whatever is called by operations.save() could be in a different thread.
+		// note that this will also clear out Options, but that's ok as any options would not work
+		// with all of insert/upsert/replace. If Options are needed, use template.insertById/upsertById/replaceById
+		getReactiveTemplate().setPseudoArgs(null);
+		return Streamable.of(entities).stream().map((e) -> operations.save(e,scopeName, collectionName)).collect(StreamUtils.toUnmodifiableList());
 	}
 
 	@Override

--- a/src/main/java/org/springframework/data/couchbase/repository/support/SimpleReactiveCouchbaseRepository.java
+++ b/src/main/java/org/springframework/data/couchbase/repository/support/SimpleReactiveCouchbaseRepository.java
@@ -77,7 +77,7 @@ public class SimpleReactiveCouchbaseRepository<T, ID> extends CouchbaseRepositor
 		String scopeName = getScope();
 		String collectionName = getCollection();
 		// clear out the PseudoArgs here as whatever is called by operations.save() could be in a different thread.
-		// not that this will also clear out Options, but that's ok as any options would not work
+		// note that this will also clear out Options, but that's ok as any options would not work
 		// with all of insert/upsert/replace. If Options are needed, use template.insertById/upsertById/replaceById
 		getReactiveTemplate().setPseudoArgs(null);
 		return operations.save(entity, scopeName, collectionName);
@@ -88,6 +88,10 @@ public class SimpleReactiveCouchbaseRepository<T, ID> extends CouchbaseRepositor
 		Assert.notNull(entities, "The given Iterable of entities must not be null!");
 		String scope = getScope();
 		String collection = getCollection();
+		// clear out the PseudoArgs here as whatever is called by operations.save() could be in a different thread.
+		// note that this will also clear out Options, but that's ok as any options would not work
+		// with all of insert/upsert/replace. If Options are needed, use template.insertById/upsertById/replaceById
+		getReactiveTemplate().setPseudoArgs(null);
 		return Flux.fromIterable(entities).flatMap(e -> save(e, scope, collection));
 	}
 
@@ -96,6 +100,10 @@ public class SimpleReactiveCouchbaseRepository<T, ID> extends CouchbaseRepositor
 		Assert.notNull(entityStream, "The given Iterable of entities must not be null!");
 		String scope = getScope();
 		String collection = getCollection();
+		// clear out the PseudoArgs here as whatever is called by operations.save() could be in a different thread.
+		// note that this will also clear out Options, but that's ok as any options would not work
+		// with all of insert/upsert/replace. If Options are needed, use template.insertById/upsertById/replaceById
+		getReactiveTemplate().setPseudoArgs(null);
 		return Flux.from(entityStream).flatMap(e -> save(e, scope, collection));
 	}
 

--- a/src/test/java/org/springframework/data/couchbase/repository/query/CouchbaseRepositoryQueryCollectionIntegrationTests.java
+++ b/src/test/java/org/springframework/data/couchbase/repository/query/CouchbaseRepositoryQueryCollectionIntegrationTests.java
@@ -17,6 +17,8 @@ package org.springframework.data.couchbase.repository.query;
 
 import static com.couchbase.client.core.io.CollectionIdentifier.DEFAULT_SCOPE;
 import static com.couchbase.client.java.query.QueryScanConsistency.REQUEST_PLUS;
+import static java.util.Arrays.asList;
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
@@ -443,4 +445,23 @@ public class CouchbaseRepositoryQueryCollectionIntegrationTests extends Collecti
 		}
 	}
 
+
+	@Test	// DATACOUCH-650, SDC-1939
+	void deleteAllById() {
+
+		Airport vienna = new Airport("airports::vie", "vie", "LOWW");
+		Airport frankfurt = new Airport("airports::fra", "fra", "EDDZ");
+		Airport losAngeles = new Airport("airports::lax", "lax", "KLAX");
+		AirportRepository ar = airportRepository.withScope(scopeName).withCollection(collectionName);
+		try {
+			ar.saveAll(asList(vienna, frankfurt, losAngeles));
+			List<Airport> airports = ar.findAllById(asList(vienna.getId(), losAngeles.getId()));
+			assertEquals(2, airports.size());
+			ar.deleteAllById(asList(vienna.getId(), losAngeles.getId()));
+			assertThat(ar.findAll()).containsExactly(frankfurt);
+			ar.deleteAll(asList(frankfurt));
+		} finally {
+			ar.deleteAll();
+		}
+	}
 }

--- a/src/test/java/org/springframework/data/couchbase/repository/query/ReactiveCouchbaseRepositoryQueryCollectionIntegrationTests.java
+++ b/src/test/java/org/springframework/data/couchbase/repository/query/ReactiveCouchbaseRepositoryQueryCollectionIntegrationTests.java
@@ -15,9 +15,12 @@
  */
 package org.springframework.data.couchbase.repository.query;
 
+import static java.util.Arrays.asList;
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
+import org.springframework.data.couchbase.domain.AirportRepository;
 import reactor.core.Disposable;
 
 import java.util.List;
@@ -295,6 +298,25 @@ public class ReactiveCouchbaseRepositoryQueryCollectionIntegrationTests extends 
 		} finally {
 			// this will fail if the above didn't use collectionName2
 			reactiveAirportRepository.withScope(scopeName).withCollection(collectionName2).deleteById(otherAirport.getId());
+		}
+	}
+
+	@Test	// DATACOUCH-650, SDC-1939
+	void deleteAllById() {
+
+		Airport vienna = new Airport("airports::vie", "vie", "LOWW");
+		Airport frankfurt = new Airport("airports::fra", "fra", "EDDZ");
+		Airport losAngeles = new Airport("airports::lax", "lax", "KLAX");
+		ReactiveAirportRepository ar = reactiveAirportRepository.withScope(scopeName).withCollection(collectionName);
+		try {
+			ar.saveAll(asList(vienna, frankfurt, losAngeles)).blockLast();
+			List<Airport> airports = ar.findAllById(asList(vienna.getId(), losAngeles.getId())).collectList().block();
+			assertEquals(2, airports.size());
+			ar.deleteAllById(asList(vienna.getId(), losAngeles.getId())).block();
+			assertThat(ar.findAll().collectList().block()).containsExactly(frankfurt);
+			ar.deleteAll(asList(frankfurt)).block();
+		} finally {
+			ar.deleteAll().block();
 		}
 	}
 

